tmpl/blog/introducing-solo5.md

```diff
@@ -0,0 +1,125 @@
+forms a base (similar to Mini-OS) from which unikernels can be built.
+It runs on fully virtualized x86 hardware (e.g., KVM/QEMU), using
+`virtio` device interfaces.
+
+Importantly, Solo5 is integrated (to some extent) with the Mirage
+toolstack, so the Solo5 version of the Mirage toolstack can build
+Mirage unikernels that run directly on KVM/QEMU instead of Xen.  As
+such, Solo5 can be considered an alternative to Mini-OS in the Mirage
+stack.  [Try it out
+today!](https://github.com/djwillia/solo5/tree/mirage)
+
+In the rest of this post, I'll give a bit of motivation about why I
+think the lowest layer of the unikernel is interesting and important,
+as well as a rough overview of the steps I took to create Solo5.
+Thanks for reading!
+
+### Why focus so far down the software stack?
+
+When people think about Mirage unikernels, one of the first things
+that comes to mind is the use of a high-level language (OCaml).
+Indeed, the Mirage community has invested lots of time and effort
+producing implementations of traditional system components (e.g., an
+entire TCP stack) in OCaml.  The pervasive use of OCaml contributes to
+security arguments for Mirage unikernels (strong type systems are
+good) and is an interesting design choice well worth exploring.
+
+But underneath all of that OCaml goodness is a little kernel layer
+written in C.  This layer has a direct impact on:
+
+* **What environments the unikernel can run on.** Mini-OS, for
+  example, assumes a paravirtualized (Xen) machine, whereas Solo5
+  targets full x86 hardware virtualization with `virtio` devices.
+
+* **Boot time.** "Hardware" initialization (or lack of it in a
+  paravirtualized case) is a major factor in achieving the 20 ms
+  unikernel boot times that are changing the way people think about
+  elasticity in the cloud.
+
+* **Memory layout and protection.** Hardware "features" like
+  page-level write protection must be exposed by the lowest layer for
+  techniques like memory tracing to be performed.  Also,
+  software-level strategies like address space layout randomization
+  require cooperation of this lowest layer.
+
+* **Low-level device interfacing.** As individual devices (e.g., NICs)
+  gain virtualization capabilities, the lowest-software layer is an
+  obvious place to interface directly with hardware.
+
+* **Threads/events.** The low-level code must ensure that device I/O
+  is asynchronous and/or fits with the higher-level synchronization
+  primitives.
+
+The most popular existing code providing this low-level kernel layer
+for is called Mini-OS.  Mini-OS was (I believe) originally written as
+a vehicle to demonstrate the paravirtualized interface offered by Xen
+for people to have a reference to port their kernels to and as a base
+for new kernel builders to build specialized Xen domains.  Mini-OS is
+a popular base for MirageOS, ClickOS, and other unikernels.  Other
+software that implements a unikernel base include rumprun and OSv.
```

Comment:
rumprun -> Rumprun (capitalised) links to http://rumpkernel.org/
OSv links to http://osv.io/